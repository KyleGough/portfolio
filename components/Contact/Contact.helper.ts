import { clsx } from 'clsx';

export const validateEmail = (email: string) => {
  const emailRegex =
    /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/;
  return !email || emailRegex.test(String(email).toLowerCase());
};

export const getFieldBorderStyle = (hasError: boolean) => {
  return clsx(
    { 'border-error': hasError },
    { 'border-link focus:border-link-hover': !hasError }
  );
};
