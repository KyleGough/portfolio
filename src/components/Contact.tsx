import React, { ReactNode, useState } from 'react';
import axios from 'axios';
import SendIcon from '../icons/SendIcon';
import RestartIcon from '../icons/RestartIcon';
import TickIcon from '../icons/TickIcon';
import FadeInWrapper from './FadeInWrapper';

function FieldError(props: { show: boolean, message: string }) {
    return (
        <span className={`${props.show ? '' : 'hidden'} transition-opactity duration-200 text-sm text-error`}>{props.message}</span>
    );
}

function Label(props: { valid: boolean, htmlFor: string, children: ReactNode }) {
    return (
        <label className={`${props.valid ? 'text-link' : 'text-error'} transition-colours duration-200 text-lg mt-8 mb-2 mr-2`} htmlFor={props.htmlFor}>{props.children}</label>
    );
}

enum EmailStatus {
    IDLE,
    SENT,
    FAIL,
    LOADING
};

export default function Contact() {
    // Form input text content.
    const [name, setName] = useState('');
    const [email, setEmail] = useState('');
    const [message, setMessage] = useState('');

    // Form input error flags.
    const [nameError, setNameError] = useState(false);
    const [emailError, setEmailError] = useState(false);
    const [emailValidError, setEmailValidError] = useState(false);
    const [messageError, setMessageError] = useState(false);   

    // Email status.
    const [status, setStatus] = useState(EmailStatus.IDLE);

    // Form field change event handlers.
    const onNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setName(e.target.value);
        setNameError(false);
    }

    const onEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setEmail(e.target.value);
        setEmailError(false);
        setEmailValidError(false);
    }

    const onMessageChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        setMessage(e.target.value);
        setMessageError(false);
    }

    const validateEmail = (email: string) => {
        //eslint-disable-next-line
        const res = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return !email || res.test(String(email).toLowerCase());
    }

    const isNotEmpty = (field: string) => {
        return Boolean(field);
    }
  
    const sendMessage = (e: React.MouseEvent<HTMLButtonElement>) => {
        // No action if another request has been made.
        if (status === EmailStatus.LOADING || status === EmailStatus.SENT) {        
            e.currentTarget.blur();
            return;
        }
              
        setStatus(EmailStatus.LOADING);  

        // Form validation.
        let validForm = true;
        
        const validator = (condition: boolean, setError: (a: boolean) => void) => {
            validForm = validForm && condition;
            setError(!condition);
        }

        validator(isNotEmpty(name), setNameError);
        validator(isNotEmpty(email), setEmailError);
        validator(validateEmail(email), setEmailValidError);
        validator(isNotEmpty(message), setMessageError);
        
        if (!validForm) {           
            setStatus(EmailStatus.IDLE);
            e.currentTarget.blur();
            return;
        }
    
        // Send Email.
        axios.post('/api/sendMessage', {
            name: name,
            email: email,
            message: message
        }).then((res) => {
            console.log('Email Sent');
            setStatus(EmailStatus.SENT);
        }).catch((error) => {   
            setStatus(EmailStatus.FAIL);
        });       
        
        // Unfocus element.
        e.currentTarget.blur();
    }

    return (
        <div className='flex justify-center'>
        <fieldset>
            <legend className='font-thin text-6xl text-center w-full mb-16 text-nav-light'>Contact Me</legend>
          
            <FadeInWrapper>
                <Label valid={!nameError} htmlFor='name'>Name</Label>
                <FieldError show={nameError} message='Required!' />
                <input
                    onChange={onNameChange}
                    className={`${nameError ? 'border-error' : 'border-link focus:border-link-hover'} 
                        transition-colours duration-200 outline-none mt-2 mb-8 block px-4 py-2
                        w-full xs:w-field border-2 rounded-2xl shadow outline-1 caret-link-hover`}
                    type='email'
                    id='name'
                    name='name'
                    placeholder='Your name here'
                    required
                    maxLength={50}
                    readOnly={status === EmailStatus.LOADING || status === EmailStatus.SENT}
                />
            </FadeInWrapper>
        
            <FadeInWrapper>
                <Label valid={!(emailError || emailValidError)} htmlFor='email'>Email</Label>
                <FieldError show={emailError} message='Required!' />
                <FieldError show={emailValidError} message='Invalid Email!' />
                <input
                    onChange={onEmailChange}
                    className={`${emailError || emailValidError ? 'border-error' : 'border-link focus:border-link-hover'} 
                        transition-colours duration-200 outline-none mt-2 mb-8 block px-4 py-2
                        w-full xs:w-field border-2 rounded-2xl shadow outline-1 caret-link-hover`}
                    type='text'
                    id='email'
                    name='email'
                    placeholder='Your email here'
                    required
                    maxLength={254}
                    readOnly={status === EmailStatus.LOADING || status === EmailStatus.SENT}
                />
            </FadeInWrapper>
        
            <FadeInWrapper>
                <Label valid={!messageError} htmlFor='message'>Message</Label>
                <FieldError show={messageError} message='Required!' />
                <textarea
                    onChange={onMessageChange}
                    className={`${messageError ? 'border-error' : 'border-link focus:border-link-hover'} 
                        transition-colours duration-200 mt-2 mb-8 block resize-none px-4 py-2
                        border-2 border-link focus:border-link-hover rounded-2xl shadow
                        w-full xs:w-field outline-none caret-link-hover max-h-48 h-48`}
                    id='message'
                    name='message'
                    placeholder='Enter your message'
                    required
                    maxLength={1024}
                    readOnly={status === EmailStatus.LOADING || status === EmailStatus.SENT}
                />
            </FadeInWrapper>
            
            <FadeInWrapper>
                <button
                    onClick={sendMessage}
                    className={
                        `${status === EmailStatus.SENT ?
                            'text-link-hover border-link-hover cursor-default' 
                        : status === EmailStatus.LOADING ?
                            'text-disabled border-link-disabled cursor-default'
                        : status === EmailStatus.FAIL ?
                            'text-error border-error'
                        : 'text-link hover:text-link-hover focus:text-link-hover border-link hover:border-link-hover focus:border-link-hover'
                        }
                        group my-8 shimmer group flex items-center px-12 py-4 bg-background 
                        rounded-full border-2 shadow mx-auto`
                    }
                    type='submit'
                >
                    {status === EmailStatus.SENT ? 'Message Sent' : status === EmailStatus.FAIL ? 'Message Fail' : 'Send Message'}
                    {status === EmailStatus.IDLE && <SendIcon className='ml-4 w-6 h-6 fill-link group-hover:fill-link-hover group-focus:fill-link-hover' />}
                    {status === EmailStatus.SENT && <TickIcon className='ml-4 w-6 h-6 fill-link-hover' />}
                    {status === EmailStatus.FAIL && <RestartIcon className='ml-4 w-6 h-6 fill-error ' />}
                    {status === EmailStatus.LOADING && <RestartIcon className='ml-4 w-6 h-6 fill-disabled animate-spin' />}
                </button>
            </FadeInWrapper>
        </fieldset>
        </div>
    );
}

