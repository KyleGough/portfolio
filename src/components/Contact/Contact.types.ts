export enum EmailStatus {
  IDLE,
  SENT,
  FAIL,
  LOADING,
}
