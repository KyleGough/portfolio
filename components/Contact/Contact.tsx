import { FadeIn } from '@components/FadeIn';
import { Endpoints } from '@utilities/endpoints';
import axios from 'axios';
import { clsx } from 'clsx';
import React, { useState } from 'react';

import { getFieldBorderStyle, validateEmail } from './Contact.helper';
import { EmailStatus } from './Contact.types';
import { ContactFieldError } from './ContactFieldError';
import { ContactLabel } from './ContactLabel';
import { ContactSendButton } from './ContactSendButton';

export const Contact: React.FC = () => {
  // Form input text content.
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [message, setMessage] = useState('');

  // Form input error flags.
  const [nameError, setNameError] = useState(false);
  const [emailError, setEmailError] = useState(false);
  const [emailValidError, setEmailValidError] = useState(false);
  const [messageError, setMessageError] = useState(false);

  // Email status.
  const [status, setStatus] = useState(EmailStatus.IDLE);

  // Form field change event handlers.
  const onNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setName(event.target.value);
    setNameError(false);
  };

  const onEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
    setEmailError(false);
    setEmailValidError(false);
  };

  const onMessageChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setMessage(event.target.value);
    setMessageError(false);
  };

  const sendMessage = (event: React.MouseEvent<HTMLButtonElement>) => {
    // No action if another request has been made.
    if (status === EmailStatus.LOADING || status === EmailStatus.SENT) {
      event.currentTarget.blur();
      return;
    }

    setStatus(EmailStatus.LOADING);

    // Form validation.
    let validForm = true;

    const validator = (condition: boolean, setError: (a: boolean) => void) => {
      validForm = validForm && condition;
      setError(!condition);
    };

    validator(!!name, setNameError);
    validator(!!email, setEmailError);
    validator(validateEmail(email), setEmailValidError);
    validator(!!message, setMessageError);

    if (!validForm) {
      setStatus(EmailStatus.IDLE);
      event.currentTarget.blur();
      return;
    }

    // Send Email.
    axios
      .post(Endpoints.CONTACT, {
        name: name,
        email: email,
        message: message,
      })
      .then(() => {
        setStatus(EmailStatus.SENT);
      })
      .catch(() => {
        setStatus(EmailStatus.FAIL);
      });

    // Unfocus element.
    event.currentTarget.blur();
  };

  const inputCommonStyles = clsx(
    'block shadow',
    'transition-colours duration-200',
    'mt-2 mb-8 px-4 py-2',
    'w-full xs:w-field',
    'border-2 rounded-2xl',
    'outline-none caret-link-hover bg-background'
  );

  return (
    <form className="flex justify-center">
      <fieldset>
        <legend className="font-thin text-6xl text-center w-full mb-16 text-nav-light">
          Contact Me
        </legend>

        <FadeIn>
          <ContactLabel valid={!nameError} htmlFor="name">
            Name
          </ContactLabel>
          {nameError && <ContactFieldError message="Required!" />}
          <input
            onChange={onNameChange}
            className={clsx(getFieldBorderStyle(nameError), inputCommonStyles)}
            type="text"
            id="name"
            name="name"
            placeholder="Your name here"
            required
            maxLength={50}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactLabel
            valid={!(emailError || emailValidError)}
            htmlFor="email"
          >
            Email
          </ContactLabel>
          {emailError && <ContactFieldError message="Required!" />}
          {emailValidError && <ContactFieldError message="Invalid Email!" />}
          <input
            onChange={onEmailChange}
            className={clsx(
              getFieldBorderStyle(emailError || emailValidError),
              inputCommonStyles
            )}
            type="email"
            id="email"
            name="email"
            placeholder="Your email here"
            required
            maxLength={254}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactLabel valid={!messageError} htmlFor="message">
            Message
          </ContactLabel>
          {messageError && <ContactFieldError message="Required!" />}
          <textarea
            onChange={onMessageChange}
            className={clsx(
              getFieldBorderStyle(messageError),
              inputCommonStyles,
              'max-h-48',
              'h-48',
              'resize-none'
            )}
            id="message"
            name="message"
            placeholder="Enter your message"
            required
            maxLength={1024}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactSendButton onClick={sendMessage} status={status} />
        </FadeIn>
      </fieldset>
    </form>
  );
};
