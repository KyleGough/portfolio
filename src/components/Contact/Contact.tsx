import axios from 'axios';
import React, { useState } from 'react';

import { FadeIn } from '../FadeIn';
import { getFieldBorderStyle,validateEmail } from './Contact.helper';
import { EmailStatus } from './Contact.types';
import { ContactFieldError } from './ContactFieldError';
import { ContactLabel } from './ContactLabel';
import { ContactSendButton } from './ContactSendButton';

export const Contact: React.FC = () => {
  // Form input text content.
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [message, setMessage] = useState('');

  // Form input error flags.
  const [nameError, setNameError] = useState(false);
  const [emailError, setEmailError] = useState(false);
  const [emailValidError, setEmailValidError] = useState(false);
  const [messageError, setMessageError] = useState(false);

  // Email status.
  const [status, setStatus] = useState(EmailStatus.IDLE);

  // Form field change event handlers.
  const onNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setName(event.target.value);
    setNameError(false);
  };

  const onEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
    setEmailError(false);
    setEmailValidError(false);
  };

  const onMessageChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setMessage(event.target.value);
    setMessageError(false);
  };

  const sendMessage = (event: React.MouseEvent<HTMLButtonElement>) => {
    // No action if another request has been made.
    if (status === EmailStatus.LOADING || status === EmailStatus.SENT) {
      event.currentTarget.blur();
      return;
    }

    setStatus(EmailStatus.LOADING);

    // Form validation.
    let validForm = true;

    const validator = (condition: boolean, setError: (a: boolean) => void) => {
      validForm = validForm && condition;
      setError(!condition);
    };

    validator(Boolean(name), setNameError);
    validator(Boolean(email), setEmailError);
    validator(validateEmail(email), setEmailValidError);
    validator(Boolean(message), setMessageError);

    if (!validForm) {
      setStatus(EmailStatus.IDLE);
      event.currentTarget.blur();
      return;
    }

    // Send Email.
    axios
      .post('/api/sendMessage', {
        name: name,
        email: email,
        message: message,
      })
      .then(() => {
        setStatus(EmailStatus.SENT);
      })
      .catch(() => {
        setStatus(EmailStatus.FAIL);
      });

    // Unfocus element.
    event.currentTarget.blur();
  };

  return (
    <div className="flex justify-center">
      <fieldset>
        <legend className="font-thin text-6xl text-center w-full mb-16 text-nav-light">
          Contact Me
        </legend>

        <FadeIn>
          <ContactLabel valid={!nameError} htmlFor="name">
            Name
          </ContactLabel>
          <ContactFieldError show={nameError} message="Required!" />
          <input
            onChange={onNameChange}
            className={`${getFieldBorderStyle(nameError)}
                        transition-colours duration-200 outline-none mt-2 mb-8 block px-4 py-2
                        w-full xs:w-field border-2 rounded-2xl shadow outline-1 caret-link-hover bg-background`}
            type="email"
            id="name"
            name="name"
            placeholder="Your name here"
            required
            maxLength={50}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactLabel
            valid={!(emailError || emailValidError)}
            htmlFor="email"
          >
            Email
          </ContactLabel>
          <ContactFieldError show={emailError} message="Required!" />
          <ContactFieldError show={emailValidError} message="Invalid Email!" />
          <input
            onChange={onEmailChange}
            className={`${getFieldBorderStyle(emailError || emailValidError)}
                        transition-colours duration-200 outline-none mt-2 mb-8 block px-4 py-2
                        w-full xs:w-field border-2 rounded-2xl shadow outline-1 caret-link-hover bg-background`}
            type="text"
            id="email"
            name="email"
            placeholder="Your email here"
            required
            maxLength={254}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactLabel valid={!messageError} htmlFor="message">
            Message
          </ContactLabel>
          <ContactFieldError show={messageError} message="Required!" />
          <textarea
            onChange={onMessageChange}
            className={`${getFieldBorderStyle(messageError)}
                        transition-colours duration-200 mt-2 mb-8 block resize-none px-4 py-2
                        border-2 border-link focus:border-link-hover rounded-2xl shadow
                        w-full xs:w-field outline-none caret-link-hover max-h-48 h-48 bg-background`}
            id="message"
            name="message"
            placeholder="Enter your message"
            required
            maxLength={1024}
            readOnly={
              status === EmailStatus.LOADING || status === EmailStatus.SENT
            }
          />
        </FadeIn>

        <FadeIn>
          <ContactSendButton onClick={sendMessage} status={status} />
        </FadeIn>
      </fieldset>
    </div>
  );
};
