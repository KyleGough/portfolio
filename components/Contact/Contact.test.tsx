import { mockIntersectionObserver } from '@mocks/mockIntersectionObserver';
import { act, fireEvent, render, screen } from '@testing-library/react';
import axios from 'axios';
import React from 'react';

import { Contact } from './Contact';

describe('Contact component', () => {
  beforeEach(mockIntersectionObserver);

  it('renders', async () => {
    const { container } = render(<Contact />);
    expect(container).toMatchSnapshot();
  });

  it('contact from gives errors when required fields are empty', async () => {
    const spy = jest.spyOn(axios, 'post');
    render(<Contact />);

    // Send message.
    const sendBtn = screen.getByRole('button');
    act(() => {
      fireEvent.click(sendBtn);
    });

    // All 3 fields should show required error.
    expect(screen.getAllByText('Required!')).toHaveLength(3);

    // Email API should not have been called.
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it('contact from gives error when given invalid email', async () => {
    const spy = jest.spyOn(axios, 'post');
    render(<Contact />);

    // Update email field.
    const emailField = screen.getByPlaceholderText('Your email here');
    act(() => {
      fireEvent.change(emailField, { target: { value: 'foo' } });
    });

    // Send message.
    const sendBtn = screen.getByRole('button');
    act(() => {
      fireEvent.click(sendBtn);
    });

    // Email field should show invalid email error.
    expect(screen.getByText('Invalid Email!')).toBeVisible();

    // Name and message fields should show required error.
    expect(screen.getAllByText('Required!')).toHaveLength(2);

    // Email API should not have been called.
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it('contact form submits on valid inputs', async () => {
    const axiosMock = jest
      .spyOn(axios, 'post')
      .mockReturnValue(Promise.reject(() => jest.fn().mockReturnValue('test')));

    render(<Contact />);

    const nameField = screen.getByPlaceholderText('Your name here');
    const emailField = screen.getByPlaceholderText('Your email here');
    const messageField = screen.getByPlaceholderText('Enter your message');

    // Update all form fields.
    act(() => {
      fireEvent.change(nameField, { target: { value: 'Kyle Gough' } });
    });

    act(() => {
      fireEvent.change(emailField, {
        target: { value: 'foo.bar@baz.com' },
      });
    });

    act(() => {
      fireEvent.change(messageField, {
        target: { value: 'Custom Message' },
      });
    });

    // Send message.
    const sendBtn = screen.getByRole('button');
    await act(async () => {
      fireEvent.click(sendBtn);
    });

    // Email API should have been called.
    expect(axiosMock).toHaveBeenCalledTimes(1);
    expect(axiosMock).toHaveBeenCalledWith('/api/sendMessage', {
      name: 'Kyle Gough',
      email: 'foo.bar@baz.com',
      message: 'Custom Message',
    });
  });
});
