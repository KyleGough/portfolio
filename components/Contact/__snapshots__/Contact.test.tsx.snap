// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component renders 1`] = `
<div>
  <form
    class="flex justify-center"
  >
    <fieldset>
      <legend
        class="font-thin text-6xl text-center w-full mb-16 text-nav-light"
      >
        Contact Me
      </legend>
      <div
        class="opacity-0 transition-opacity duration-1000"
      >
        <label
          class="text-link transition-colours duration-200 text-lg mt-8 mb-2 mr-2"
          for="name"
        >
          Name
        </label>
        <input
          class="border-link focus:border-link-hover block shadow transition-colours duration-200 mt-2 mb-8 px-4 py-2 w-full xs:w-field border-2 rounded-2xl outline-none caret-link-hover bg-background"
          id="name"
          maxlength="50"
          name="name"
          placeholder="Your name here"
          required=""
          type="text"
        />
      </div>
      <div
        class="opacity-0 transition-opacity duration-1000"
      >
        <label
          class="text-link transition-colours duration-200 text-lg mt-8 mb-2 mr-2"
          for="email"
        >
          Email
        </label>
        <input
          class="border-link focus:border-link-hover block shadow transition-colours duration-200 mt-2 mb-8 px-4 py-2 w-full xs:w-field border-2 rounded-2xl outline-none caret-link-hover bg-background"
          id="email"
          maxlength="254"
          name="email"
          placeholder="Your email here"
          required=""
          type="email"
        />
      </div>
      <div
        class="opacity-0 transition-opacity duration-1000"
      >
        <label
          class="text-link transition-colours duration-200 text-lg mt-8 mb-2 mr-2"
          for="message"
        >
          Message
        </label>
        <textarea
          class="border-link focus:border-link-hover block shadow transition-colours duration-200 mt-2 mb-8 px-4 py-2 w-full xs:w-field border-2 rounded-2xl outline-none caret-link-hover bg-background max-h-48 h-48 resize-none"
          id="message"
          maxlength="1024"
          name="message"
          placeholder="Enter your message"
          required=""
        />
      </div>
      <div
        class="opacity-0 transition-opacity duration-1000"
      >
        <button
          class="text-link hover:text-link-hover focus:text-link-hover border-link hover:border-link-hover focus:border-link-hover group flex items-center mx-auto my-8 px-12 py-4 bg-background rounded-full border-2 shimmer shadow"
          type="button"
        >
          Send Message
          <svg
            class="ml-4 w-6 h-6 fill-link group-hover:fill-link-hover group-focus:fill-link-hover"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
            />
          </svg>
        </button>
      </div>
    </fieldset>
  </form>
</div>
`;
